#### **PS - V**
#### **DONE BY CS20B1057 A S V DHANUSH**

In [85]:
#importing libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [86]:
#import the csv avocado file
avocado = pd.read_csv("avocado.csv")
avocado.head()


Date AveragePrice  Total Volume     4046       4225    4770  Total Bags  Small Bags  Large Bags  XLarge Bags          type  year  region
0  27-12-2015         1.33      64236.62  1036.74   54454.85   48.16     8696.87     8603.62       93.25          0.0  conventional  2015  Albany
1  20-12-2015         1.35      54876.98   674.28   44638.81   58.33     9505.56     9408.07       97.49          0.0  conventional  2015  Albany
2  13-12-2015         0.93     118220.22   794.70  109149.67  130.50     8145.35     8042.21      103.14          0.0  conventional  2015  Albany
3  06-12-2015         1.08      78992.15  1132.00   71976.41   72.58     5811.16     5677.40      133.76          0.0  conventional  2015  Albany
4  29-11-2015         1.29      51039.60   941.48   43838.39   75.78     6183.95     5986.26      197.69          0.0  conventional  2015  Albany

In [87]:
avocado_shape = avocado.shape
avocado_shape
avocado.columns

Index(['Date', 'AveragePrice', 'Total Volume', '4046', '4225', '4770',
       'Total Bags', 'Small Bags', 'Large Bags', 'XLarge Bags', 'type', 'year',
       'region'],
      dtype='object')

Q1. Select a subset of relevant attributes from the given dataset that are necessary to know about the total volume of
avocados with product lookup codes (PLU) 4046, 4225, 4770) which are of organic type. (Use AVOCADO dataset)

In [88]:
import pandas as pd

plu_data = avocado[(avocado['type'] == 'organic') &
                             ((avocado['4046'] == avocado['4046']) |
                              (avocado['4225'] == avocado['4225'] ) |
                              (avocado['4770'] == avocado['4770']))]
#print(plu_data)
# Calculate the total volume by summing the 'Total Volume' column
total_volume = plu_data['Total Volume'].sum()

print("Total Volume of avocados with PLU 4046, 4225, and 4770 And are orgainc:", total_volume)


Total Volume of avocados with PLU 4046, 4225, and 4770 And are orgainc: 436197578.46999997


##Total Volume of avocados with PLU 4046, 4225, and 4770 And are orgainc : 436197578.47

Q2. Discard all duplicate entries in the dataset given and fill all the missing values in the attribute “AveragePrice” as 1.25. Also print the size of the dataset before and after removing duplicates. (Use Trail dataset)

In [89]:
avocado_null = avocado.isnull().sum()
print("\nList of Attributes and correspondning number of NaN values\n",avocado_null)


List of Attributes and correspondning number of NaN values
 Date             0
AveragePrice    28
Total Volume     0
4046             0
4225             0
4770             0
Total Bags       0
Small Bags       0
Large Bags       0
XLarge Bags      0
type             0
year             0
region           0
dtype: int64


###There are 28 NaN values for the AveragePrice Column

In [90]:
print("Shape of dataset before dropping duplicates : ",avocado.shape)

avocado_no_dup = avocado.drop_duplicates()#droping duplicates
avocado_without_dup_shape = avocado_no_dup.shape
print("Shape of dataset after dropping duplicates : ",avocado_no_dup.shape)

avocado_without_na = avocado_no_dup.fillna(1.25) #giving 1.25 for NaN values

#print(avocado_without_na)
print("\nList of Attributes and correspondning number of NaN values\n",avocado_without_na.isnull().sum())

Shape of dataset before dropping duplicates :  (18250, 13)
Shape of dataset after dropping duplicates :  (18249, 13)

List of Attributes and correspondning number of NaN values
 Date            0
AveragePrice    0
Total Volume    0
4046            0
4225            0
4770            0
Total Bags      0
Small Bags      0
Large Bags      0
XLarge Bags     0
type            0
year            0
region          0
dtype: int64


In [91]:
data = avocado_without_na
data.tail()

Date AveragePrice  Total Volume     4046     4225    4770  Total Bags  Small Bags  Large Bags  XLarge Bags     type  year            region
18244  04-02-2018         1.63      17074.83  2046.96  1529.20    0.00    13498.67    13066.82      431.85          0.0  organic  2018  WestTexNewMexico
18245  28-01-2018         1.71      13888.04  1191.70  3431.50    0.00     9264.84     8940.04      324.80          0.0  organic  2018  WestTexNewMexico
18246  21-01-2018         1.87      13766.76  1191.92  2452.79  727.94     9394.11     9351.80       42.31          0.0  organic  2018  WestTexNewMexico
18247  14-01-2018         1.93      16205.22  1527.63  2981.04  727.01    10969.54    10919.54       50.00          0.0  organic  2018  WestTexNewMexico
18248  07-01-2018         1.62      17489.58  2894.77  2356.13  224.53    12014.15    11988.14       26.01          0.0  organic  2018  WestTexNewMexico

In [92]:
#storing the dataframe in new df
new_data = data.copy(deep = False)

####There is 1 duplicate in this dataset

Q3. Binarize the attribute “Year”. Set the threshold above 2016 and print it without truncation. (Use AVOCADO
dataset)

In [93]:
# Setting 2016 aws threshold
threshold = 2016

# Binarizing the "year" attribute (0 or 1)
new_data['year'] = (data['year'] > threshold).astype(int)

new_data.tail()


Date AveragePrice  Total Volume     4046     4225    4770  Total Bags  Small Bags  Large Bags  XLarge Bags     type  year            region
18244  04-02-2018         1.63      17074.83  2046.96  1529.20    0.00    13498.67    13066.82      431.85          0.0  organic     1  WestTexNewMexico
18245  28-01-2018         1.71      13888.04  1191.70  3431.50    0.00     9264.84     8940.04      324.80          0.0  organic     1  WestTexNewMexico
18246  21-01-2018         1.87      13766.76  1191.92  2452.79  727.94     9394.11     9351.80       42.31          0.0  organic     1  WestTexNewMexico
18247  14-01-2018         1.93      16205.22  1527.63  2981.04  727.01    10969.54    10919.54       50.00          0.0  organic     1  WestTexNewMexico
18248  07-01-2018         1.62      17489.58  2894.77  2356.13  224.53    12014.15    11988.14       26.01          0.0  organic     1  WestTexNewMexico

Q4. Transform all categorical attributes in the dataset AVOCADO using Integer Encoding.

In [94]:
# Defining a list of categorical columns based on your attributes
categorical_columns = ['type', 'region']
operation1_data = new_data.copy(deep=False)
# Performing Integer Encoding for each categorical column
for column in categorical_columns:
    operation1_data[column] = new_data[column].astype('category').cat.codes


operation1_data.tail()


Date AveragePrice  Total Volume     4046     4225    4770  Total Bags  Small Bags  Large Bags  XLarge Bags  type  year  region
18244  04-02-2018         1.63      17074.83  2046.96  1529.20    0.00    13498.67    13066.82      431.85          0.0     1     1      53
18245  28-01-2018         1.71      13888.04  1191.70  3431.50    0.00     9264.84     8940.04      324.80          0.0     1     1      53
18246  21-01-2018         1.87      13766.76  1191.92  2452.79  727.94     9394.11     9351.80       42.31          0.0     1     1      53
18247  14-01-2018         1.93      16205.22  1527.63  2981.04  727.01    10969.54    10919.54       50.00          0.0     1     1      53
18248  07-01-2018         1.62      17489.58  2894.77  2356.13  224.53    12014.15    11988.14       26.01          0.0     1     1      53

Q5. Transform the attribute = “Region” in the given dataset AVOCADO using One-Hot Encoding.

In [97]:
operation2_data = new_data.copy(deep = False)

# Performing One-Hot Encoding on the 'region' attribute
avocado_data_encoded = pd.get_dummies(operation2_data, columns=['region'], prefix=['Region'])

avocado_data_encoded.head()

Date AveragePrice  Total Volume     4046       4225    4770  Total Bags  Small Bags  Large Bags  XLarge Bags          type  year  Region_Albany  Region_Atlanta  Region_BaltimoreWashington  Region_Boise  Region_Boston  Region_BuffaloRochester  Region_California  Region_Charlotte  Region_Chicago  Region_CincinnatiDayton  Region_Columbus  Region_DallasFtWorth  Region_Denver  Region_Detroit  Region_GrandRapids  Region_GreatLakes  Region_HarrisburgScranton  Region_HartfordSpringfield  Region_Houston  Region_Indianapolis  Region_Jacksonville  Region_LasVegas  Region_LosAngeles  Region_Louisville  Region_MiamiFtLauderdale  Region_Midsouth  Region_Nashville  Region_NewOrleansMobile  Region_NewYork  Region_Northeast  Region_NorthernNewEngland  Region_Orlando  Region_Philadelphia  Region_PhoenixTucson  Region_Pittsburgh  Region_Plains  Region_Portland  Region_RaleighGreensboro  Region_RichmondNorfolk  Region_Roanoke  Region_Sacramento  Region_SanDiego  Region_SanFrancisco  Region_Seattle  Region_SouthCarolina  Region_SouthCentral  Region_Southeast  Region_Spokane  Region_StLouis  Region_Syracuse  Region_Tampa  Region_TotalUS  Region_West  Region_WestTexNewMexico
0  27-12-2015         1.33      64236.62  1036.74   54454.85   48.16     8696.87     8603.62       93.25          0.0  conventional     0              1               0                           0             0              0                        0                  0                 0               0                        0                0                     0              0               0                   0                  0                          0                           0               0                    0                    0                0                  0                  0                         0                0                 0                        0               0                 0                          0               0                    0                     0                  0              0                0                         0                       0               0                  0                0                    0               0                     0                    0                 0               0               0                0             0               0            0                        0
1  20-12-2015         1.35      54876.98   674.28   44638.81   58.33     9505.56     9408.07       97.49          0.0  conventional     0              1               0                           0             0              0                        0                  0                 0               0                        0                0                     0              0               0                   0                  0                          0                           0               0                    0                    0                0                  0                  0                         0                0                 0                        0               0                 0                          0               0                    0                     0                  0              0                0                         0                       0               0                  0                0                    0               0                     0                    0                 0               0               0                0             0               0            0                        0
2  13-12-2015         0.93     118220.22   794.70  109149.67  130.50     8145.35     8042.21      103.14          0.0  conventional     0              1               0                           0             0              0                        0                  0                 0               0                        0                0                     0              0               0                   0                  0                          0     

Q6. Ignore the tuples that hold missing values and print the subset of data from AVOCADO dataset excluding “NaN”
values.

In [100]:
avocado_data = pd.read_csv('avocado.csv')  # Replace with the actual file path

# Removing rows with missing values (NaN) and creaitng a subset dataframe
avocado_data_subset = avocado_data.dropna()

avocado_data_subset.head()


Date AveragePrice  Total Volume     4046       4225    4770  Total Bags  Small Bags  Large Bags  XLarge Bags          type  year  region
0  27-12-2015         1.33      64236.62  1036.74   54454.85   48.16     8696.87     8603.62       93.25          0.0  conventional  2015  Albany
1  20-12-2015         1.35      54876.98   674.28   44638.81   58.33     9505.56     9408.07       97.49          0.0  conventional  2015  Albany
2  13-12-2015         0.93     118220.22   794.70  109149.67  130.50     8145.35     8042.21      103.14          0.0  conventional  2015  Albany
3  06-12-2015         1.08      78992.15  1132.00   71976.41   72.58     5811.16     5677.40      133.76          0.0  conventional  2015  Albany
4  29-11-2015         1.29      51039.60   941.48   43838.39   75.78     6183.95     5986.26      197.69          0.0  conventional  2015  Albany

In [101]:
print("Shape Before Dropiing NaN : ",avocado_data.shape)
print("Shape After Dropiing NaN : ",avocado_data_subset.shape)

Shape Before Dropiing NaN :  (18250, 13)
Shape After Dropiing NaN :  (18222, 13)


####28 Entries are dropped

Q7. Drop the attribute that has high nullity as it facilitates efficient prediction. (Use AVOCADO dataset)

In [123]:
_nan = avocado_data.isna().sum()
dict_nan = dict(_nan)
for i in dict_nan.items():
  print(i)
#removing the AveragePrice Column as it has most number of NaN values

df_new = avocado_data.drop(['AveragePrice'],axis =1)

('Date', 0)
('AveragePrice', 28)
('Total Volume', 0)
('4046', 0)
('4225', 0)
('4770', 0)
('Total Bags', 0)
('Small Bags', 0)
('Large Bags', 0)
('XLarge Bags', 0)
('type', 0)
('year', 0)
('region', 0)


In [126]:
df_new.shape #it has been dropped

(18250, 12)

Q8. Study the entire dataset and report the complete statistical summary about the data (Use AVOCADO dataset)

● Dimension of the dataset

● Most frequently occurring value under every attribute.

● Datatype of every attribute

● Count

● Mean

● Standard Deviation

● Minimum Value

● Maximum value

● 25% o

● Median i.e. 50%

● 75%

● Find whether the class distribution of dataset is
imbalanced. (Note: Fix the class label as “Type” in the
given dataset)

● Correlation matrix

● Skewness of every attribute.

In [128]:
avocado  = pd.read_csv('avocado.csv')

dimension = avocado_data.shape

# Most frequently occurring value under every attribute
most_frequent_values = avocado_data.mode().iloc[0]

# Datatype of every attribute
data_types = avocado_data.dtypes

# Count, Mean, Std, Min, Max, 25%, Median, 75% are present in describe
summary_statistics = avocado_data.describe()

# Class distribution of the dataset
class_distribution = avocado_data['type'].value_counts()

# Correlation matrix
correlation_matrix = avocado_data.corr()

# Skewness of every attribute
skewness = avocado_data.skew()

print("Dimension of the dataset:", dimension)

print("\nMost frequently occurring value under every attribute:\n", most_frequent_values)

print("\nDatatype of every attribute:\n", data_types)

print("\nCount, Mean, Std, Min, Max, 25%, Median, 75%:\n", summary_statistics)

print("\nClass distribution of the dataset (Type):\n", class_distribution)

print("\nCorrelation matrix:\n", correlation_matrix)

print("\nSkewness of every attribute:", skewness)


Dimension of the dataset: (18250, 13)

Most frequently occurring value under every attribute:
 Date              18-03-2018
AveragePrice            1.15
Total Volume         2038.99
4046                     0.0
4225                     0.0
4770                     0.0
Total Bags               0.0
Small Bags               0.0
Large Bags               0.0
XLarge Bags              0.0
type            conventional
year                  2017.0
region                Albany
Name: 0, dtype: object

Datatype of every attribute:
 Date             object
AveragePrice     object
Total Volume    float64
4046            float64
4225            float64
4770            float64
Total Bags      float64
Small Bags      float64
Large Bags      float64
XLarge Bags     float64
type             object
year              int64
region           object
dtype: object

Count, Mean, Std, Min, Max, 25%, Median, 75%:
        Total Volume          4046          4225          4770    Total Bags    Small Bags    Large B

<ipython-input-128-9951f15fb4bc>:18: FutureWarning: The default value of numeric_only in DataFrame.corr is deprecated. In a future version, it will default to False. Select only valid columns or specify the value of numeric_only to silence this warning.
  correlation_matrix = avocado_data.corr()
<ipython-input-128-9951f15fb4bc>:21: FutureWarning: The default value of numeric_only in DataFrame.skew is deprecated. In a future version, it will default to False. In addition, specifying 'numeric_only=None' is deprecated. Select only valid columns or specify the value of numeric_only to silence this warning.
  skewness = avocado_data.skew()


Q9. Test drive the use of Gini Index, Information Gain, Entropy and other measures that are supported in your
platform, performing the role of data selection.

Gini Index

In [147]:
import pandas as pd
from sklearn.tree import DecisionTreeClassifier

data = pd.read_csv('adult.csv')
data_encoded = pd.get_dummies(data, columns=['workclass','education','marital-status','occupation','relationship','race','gender','native-country'])

X = data_encoded.drop(columns=['income'])  # Features
y = data_encoded['income']  # Target variable

clf = DecisionTreeClassifier()
clf.fit(X, y)

# Calculating the Gini index for each feature
gini_importance = clf.feature_importances_

for feature, importance in zip(X.columns, gini_importance):
    print(f'{feature}: {importance:.4f}')




age: 0.1097
fnlwgt: 0.1918
educational-num: 0.1092
capital-gain: 0.1057
capital-loss: 0.0397
hours-per-week: 0.0630
workclass_?: 0.0009
workclass_Federal-gov: 0.0042
workclass_Local-gov: 0.0064
workclass_Never-worked: 0.0000
workclass_Private: 0.0091
workclass_Self-emp-inc: 0.0046
workclass_Self-emp-not-inc: 0.0086
workclass_State-gov: 0.0042
workclass_Without-pay: 0.0001
education_10th: 0.0004
education_11th: 0.0005
education_12th: 0.0003
education_1st-4th: 0.0002
education_5th-6th: 0.0003
education_7th-8th: 0.0006
education_9th: 0.0005
education_Assoc-acdm: 0.0016
education_Assoc-voc: 0.0027
education_Bachelors: 0.0037
education_Doctorate: 0.0005
education_HS-grad: 0.0050
education_Masters: 0.0034
education_Preschool: 0.0001
education_Prof-school: 0.0005
education_Some-college: 0.0035
marital-status_Divorced: 0.0017
marital-status_Married-AF-spouse: 0.0004
marital-status_Married-civ-spouse: 0.1988
marital-status_Married-spouse-absent: 0.0006
marital-status_Never-married: 0.0022
marit

Entropy

In [151]:
tot_inst = len(data)
pos_inst = len(data[data['income'] == '>50K'])
neg_inst = tot_inst - pos_inst

p_positive = pos_inst / tot_inst
p_negative = neg_inst / tot_inst

entropy = -p_positive * np.log2(p_positive) - p_negative * np.log2(p_negative)

print(f'Entropy: {entropy:.4f}')

Entropy: 0.7938


Information Gain

In [152]:
information_gain = entropy #using the entropy
feature_name = 'education'
for feature_value in data[feature_name].unique():
    subset = data[data[feature_name] == feature_value]
    total_subset_instances = len(subset)
    positive_subset_instances = len(subset[subset['income'] == '>50K'])
    negative_subset_instances = total_subset_instances - positive_subset_instances
    p_positive_subset = positive_subset_instances / total_subset_instances
    p_negative_subset = negative_subset_instances / total_subset_instances
    entropy_subset = -p_positive_subset * np.log2(p_positive_subset) - p_negative_subset * np.log2(p_negative_subset)

    information_gain -= (total_subset_instances / tot_inst) * entropy_subset

# Print the Information Gain
print(f'Information Gain for feature "{feature_name}": {information_gain:.4f}')

Information Gain for feature "education": 0.0921


In [136]:
from sklearn.feature_selection import SelectKBest
from sklearn.feature_selection import chi2
from sklearn.feature_selection import mutual_info_classif
from sklearn.feature_selection import SelectFromModel
from sklearn.ensemble import RandomForestClassifier
from sklearn.datasets import load_iris

#using iris dataset
data = load_iris()
X, y = data.data, data.target

#Using Chi-squared (for classification problems)
chi2_selector = SelectKBest(chi2, k=2)  # Select the top 2 features
X_chi2_selected = chi2_selector.fit_transform(X, y)

#Using Mutual Information (for classification problems)
mi_selector = SelectKBest(mutual_info_classif, k=2)  # Select the top 2 features
X_mi_selected = mi_selector.fit_transform(X, y)

#Using Tree-based feature selection (Random Forest)
rf_selector = SelectFromModel(RandomForestClassifier(n_estimators=100))
X_rf_selected = rf_selector.fit_transform(X, y)

print("Chi-squared selected features:")
print(X_chi2_selected)
print("------------------------------")

print("\nMutual Information selected features:")
print(X_mi_selected)
print("------------------------------")

print("\nRandom Forest selected features:")
print(X_rf_selected)



Chi-squared selected features:
[[1.4 0.2]
 [1.4 0.2]
 [1.3 0.2]
 [1.5 0.2]
 [1.4 0.2]
 [1.7 0.4]
 [1.4 0.3]
 [1.5 0.2]
 [1.4 0.2]
 [1.5 0.1]
 [1.5 0.2]
 [1.6 0.2]
 [1.4 0.1]
 [1.1 0.1]
 [1.2 0.2]
 [1.5 0.4]
 [1.3 0.4]
 [1.4 0.3]
 [1.7 0.3]
 [1.5 0.3]
 [1.7 0.2]
 [1.5 0.4]
 [1.  0.2]
 [1.7 0.5]
 [1.9 0.2]
 [1.6 0.2]
 [1.6 0.4]
 [1.5 0.2]
 [1.4 0.2]
 [1.6 0.2]
 [1.6 0.2]
 [1.5 0.4]
 [1.5 0.1]
 [1.4 0.2]
 [1.5 0.2]
 [1.2 0.2]
 [1.3 0.2]
 [1.4 0.1]
 [1.3 0.2]
 [1.5 0.2]
 [1.3 0.3]
 [1.3 0.3]
 [1.3 0.2]
 [1.6 0.6]
 [1.9 0.4]
 [1.4 0.3]
 [1.6 0.2]
 [1.4 0.2]
 [1.5 0.2]
 [1.4 0.2]
 [4.7 1.4]
 [4.5 1.5]
 [4.9 1.5]
 [4.  1.3]
 [4.6 1.5]
 [4.5 1.3]
 [4.7 1.6]
 [3.3 1. ]
 [4.6 1.3]
 [3.9 1.4]
 [3.5 1. ]
 [4.2 1.5]
 [4.  1. ]
 [4.7 1.4]
 [3.6 1.3]
 [4.4 1.4]
 [4.5 1.5]
 [4.1 1. ]
 [4.5 1.5]
 [3.9 1.1]
 [4.8 1.8]
 [4.  1.3]
 [4.9 1.5]
 [4.7 1.2]
 [4.3 1.3]
 [4.4 1.4]
 [4.8 1.4]
 [5.  1.7]
 [4.5 1.5]
 [3.5 1. ]
 [3.8 1.1]
 [3.7 1. ]
 [3.9 1.2]
 [5.1 1.6]
 [4.5 1.5]
 [4.5 1.6]
 [4.7 1.5]
 [4.4 1.3]
 

##Using Adult Income Dataset for the rest of the problems

In [139]:
# Load your dataset
data = pd.read_csv('adult.csv')
data.tail()

age     workclass  fnlwgt   education  educational-num      marital-status         occupation relationship   race  gender  capital-gain  capital-loss  hours-per-week native-country income
48837   27       Private  257302  Assoc-acdm               12  Married-civ-spouse       Tech-support         Wife  White  Female             0             0              38  United-States  <=50K
48838   40       Private  154374     HS-grad                9  Married-civ-spouse  Machine-op-inspct      Husband  White    Male             0             0              40  United-States   >50K
48839   58       Private  151910     HS-grad                9             Widowed       Adm-clerical    Unmarried  White  Female             0             0              40  United-States  <=50K
48840   22       Private  201490     HS-grad                9       Never-married       Adm-clerical    Own-child  White    Male             0             0              20  United-States  <=50K
48841   52  Self-emp-inc  287927     HS-grad                9  Married-civ-spouse    Exec-managerial         Wife  White  Female         15024             0              40  United-States   >50K

In [144]:
from sklearn.preprocessing import StandardScaler
from sklearn.feature_selection import SelectKBest, f_classif

data = pd.read_csv('adult.csv')

# Data Cleaning
data = data.drop_duplicates()
data = data.dropna()

# Feature Selection
selector = SelectKBest(score_func=f_classif, k=4) #taking top 4 features
data_encoded = pd.get_dummies(data, columns=['workclass','education','marital-status','occupation','relationship','race','gender','native-country'])

# Data Transformation
X = data_encoded.drop(columns=['income'])
y = data['income']
X_selected = selector.fit_transform(X, y)


scaler = StandardScaler()
X_scaled = scaler.fit_transform(X_selected)
print(X_scaled)


[[-1.19797029 -0.92001494  1.42612294 -0.82303198]
 [-0.41976611  1.08693887 -0.70120182  1.21501961]
 [ 0.74754016  1.08693887 -0.70120182  1.21501961]
 ...
 [-0.41976611 -0.92001494 -0.70120182 -0.82303198]
 [-0.41976611 -0.92001494  1.42612294 -0.82303198]
 [-0.41976611  1.08693887 -0.70120182 -0.82303198]]
